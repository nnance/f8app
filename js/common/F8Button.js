/**
 * Copyright 2016 Facebook, Inc.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to
 * use, copy, modify, and distribute this software in source code or binary
 * form for use in connection with the web services and APIs provided by
 * Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use
 * of this software is subject to the Facebook Developer Principles and
 * Policies [http://developers.facebook.com/policy/]. This copyright notice
 * shall be included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE
 *
 * @providesModule F8Button
 * @flow
 */

import LinearGradient from 'react-native-linear-gradient';

const F8Colors = require('F8Colors');
const Image = require('Image');
const React = require('React');
const StyleSheet = require('StyleSheet');
const { Text } = require('F8Text');
const TouchableOpacity = require('TouchableOpacity');
const View = require('View');

const HEIGHT = 50;

const styles = StyleSheet.create({
  container: {
    height: HEIGHT,
    // borderRadius: HEIGHT / 2,
    // borderWidth: 1 / PixelRatio.get(),
  },
  button: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: 40,
  },
  border: {
    borderWidth: 1,
    borderColor: F8Colors.lightText,
    borderRadius: HEIGHT / 2,
  },
  primaryButton: {
    borderRadius: HEIGHT / 2,
    backgroundColor: 'transparent',
  },
  icon: {
    marginRight: 12,
  },
  caption: {
    letterSpacing: 1,
    fontSize: 12,
  },
  primaryCaption: {
    color: 'white',
  },
  secondaryCaption: {
    color: F8Colors.lightText,
  },
});

class F8Button extends React.Component {
  props: {
    type: 'primary' | 'secondary' | 'bordered';
    icon: number;
    caption: string;
    style: any;
    onPress: () => void;
  };

  render() {
    const caption = this.props.caption.toUpperCase();
    let icon;
    if (this.props.icon) {
      icon = <Image source={this.props.icon} style={styles.icon} />;
    }
    let content;
    if (this.props.type === 'primary' || this.props.type === undefined) {
      content = (
        <LinearGradient
          start={{ x: 0.5, y: 1 }} end={{ x: 1, y: 1 }}
          colors={['#6A6AD5', '#6F86D9']}
          style={[styles.button, styles.primaryButton]}
        >
          {icon}
          <Text style={[styles.caption, styles.primaryCaption]}>
            {caption}
          </Text>
        </LinearGradient>
      );
    } else {
      const border = this.props.type === 'bordered' && styles.border;
      content = (
        <View style={[styles.button, border]}>
          {icon}
          <Text style={[styles.caption, styles.secondaryCaption]}>
            {caption}
          </Text>
        </View>
      );
    }
    return (
      <TouchableOpacity
        accessibilityTraits="button"
        onPress={this.props.onPress}
        activeOpacity={0.8}
        style={[styles.container, this.props.style]}
      >
        {content}
      </TouchableOpacity>
    );
  }
}

module.exports = F8Button;
