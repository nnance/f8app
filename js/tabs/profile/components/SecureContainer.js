import React from 'react';
import {
  View,
  Text,
  TextInput,
  StyleSheet,
} from 'react-native';
import F8Button from 'F8Button';

import { colors as commonColors } from '../../../common/styles';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: commonColors.nearlyWhiteBackground,
  },
  spaceFlex: {
    flex: 1,
  },
  error: {
    margin: 10,
    alignSelf: 'center',
    color: 'red',
  },
  innerContainer: {
    flex: 8,
  },
  input: {
    flex: 1,
    height: 40,
  },
  inputBox: {
    height: 80,
    width: 300,
    padding: 20,
    borderRadius: 10,
    backgroundColor: 'rgba(255, 255, 255, 0.8)',
    borderWidth: 0.5,
    borderColor: commonColors.greyBorder,
  },
  submitBox: {
    flex: 1,
    backgroundColor: 'blue',
  },
  buttonBox: {
    padding: 60,
  },
});

class SecureContainer extends React.Component {
  constructor(...args) {
    super(...args);
    this.state = {
      passed: false,
      password: '',
      error: null,
    };
  }

  render() {
    if (this.state.passed) {
      return <View key="passed">{this.props.children}</View>;
    }
    return (<View key="secure" style={[styles.container, this.props.style]}>
      <View style={styles.spaceFlex} />
      <View style={styles.innerContainer}>
        <Text style={styles.error}>
          {this.state.error}
        </Text>
        <View style={styles.inputBox}>
          <TextInput
            onChangeText={password => this.setState({ password })}
            style={styles.input}
            placeholder="current password"
            value={this.state.password || ''}
            secureTextEntry
          />
        </View>
        <View style={styles.buttonBox}>
          <F8Button
            caption="Next" onPress={() => this.props.onCheck(this.state.password).then(() => {
              this.setState({ passed: true });
            }).catch((error) => {
              this.setState({
                error: error.message,
              });
            })}
          />
        </View>
      </View>
    </View>);
  }
}

export default SecureContainer;
