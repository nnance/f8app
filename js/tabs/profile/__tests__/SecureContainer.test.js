import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import {
  Text,
  TextInput,
} from 'react-native';
import F8Button from 'F8Button';

import SecureContainer from '../components/SecureContainer';

describe('SecureContainer', () => {
  function onCheck(password) {
    if (password === 'ok') {
      return Promise.resolve();
    }
    return Promise.reject(new Error('something wrong'));
  }

  it('render authen screen', () => {
    const tree = renderer.create(<SecureContainer><Text>secret</Text></SecureContainer>);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('not render secret', () => {
    const tree = shallow(<SecureContainer><Text>secret</Text></SecureContainer>);
    expect(tree.contains(<Text>secret</Text>)).toBe(false);
  });

  it('render secret screen', async () => {
    const tree = shallow(<SecureContainer onCheck={onCheck}><Text>secret</Text></SecureContainer>);
    tree.find(TextInput).simulate('changeText', 'ok');
    await tree.find(F8Button).props().onPress();
    expect(tree.contains(<Text>secret</Text>)).toBe(true);
  });

  it('render error', async () => {
    const tree = shallow(<SecureContainer onCheck={onCheck}><Text>secret</Text></SecureContainer>);
    tree.find(TextInput).simulate('changeText', 'fail');
    await tree.find(F8Button).props().onPress();
    expect(tree.contains('something wrong')).toBe(true);
  });
});
