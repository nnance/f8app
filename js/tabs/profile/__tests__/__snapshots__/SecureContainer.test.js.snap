exports[`SecureContainer render authen screen 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(250, 250, 250)",
        "flex": 1,
      },
      undefined,
    ]
  }>
  <View
    style={
      Object {
        "flex": 1,
      }
    } />
  <View
    style={
      Object {
        "flex": 8,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "center",
          "color": "red",
          "margin": 10,
        }
      } />
    <View
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 0.8)",
          "borderColor": "#eeeeee",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 80,
          "padding": 20,
          "width": 300,
        }
      }>
      <TextInput
        onChangeText={[Function]}
        placeholder="current password"
        secureTextEntry={true}
        style={
          Object {
            "flex": 1,
            "height": 40,
          }
        }
        value="" />
    </View>
    <View
      style={
        Object {
          "padding": 60,
        }
      }>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits="button"
        accessible={true}
        hitSlop={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "height": 50,
            "opacity": 1,
          }
        }
        testID={undefined}>
        <BVLinearGradient
          colors={
            Array [
              4285164245,
              4285499097,
            ]
          }
          end={
            Object {
              "x": 1,
              "y": 1,
            }
          }
          locations={null}
          start={
            Object {
              "x": 0.5,
              "y": 1,
            }
          }
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "paddingHorizontal": 40,
              },
              Object {
                "backgroundColor": "transparent",
                "borderRadius": 25,
              },
            ]
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "fontFamily": undefined,
                },
                Array [
                  Object {
                    "fontSize": 12,
                    "letterSpacing": 1,
                  },
                  Object {
                    "color": "white",
                  },
                ],
              ]
            }>
            NEXT
          </Text>
        </BVLinearGradient>
      </View>
    </View>
  </View>
</View>
`;
